Reimplement buttons

selbekk opened the pull request:
This pull request re-implements both `ffe-buttons` and `ffe-buttons-react` in a long-awaited major breaking change.

Please give me feedback on the way I've solved things, as I want to try to find the best possible API before merging it.

Fixes #12 and #4

selbekk committed:
feat(ffe-buttons): Simplify API

BREAKING CHANGE! This commit rewrites the ffe-buttons less package for
simplicity and ease of use.

There are now two basic button types - `base-button` and
`inline-base-button`. Each of these two button types have fewer
elements, and a few modifiers that specify which type of button it is.

Previously, the DOM structure for most buttons looked like a variation
of this:

```html
<button class="ffe-[button-type]-button">
  <span class="ffe-[button-type]-button__label">
    <span class="ffe-[button-type]-button__label-text">
        <img class="ffe-[button-type]-button__label-icon" alt="left icon" />
        Click me
        <img class="ffe-[button-type]-button__label-icon" alt="right icon" />
    </span>
    <span class="ffe-[button-type]-button__label-spinner" />
  </span>
</button>
```

For `base-button` based buttons (woah), the general DOM structure now
required is this:

```html
<button class="ffe-button ffe-button--[button-type]">
  <img class="ffe-button__icon ffe-button__icon--left" alt="left icon" />
  <span class="ffe-button__label">
    Click me
  </span>
  <img class="ffe-button__icon ffe-button__icon--right" alt="right icon" />
  <div class="ffe-button__spinner" />
</button>
```

The icons and the spinner is optional if not needed. Please refer to
the React implementation for further details on which aria-tags are
expected to be passed where, as well as which modifier flags are
available.

Note, however, how the button type only needs to be specified at the
top level. The same goes for any modifier, like loading etc. If your
button does not require a loading indicator, the entire block can be
removed without any other tweaks.

Previously, the DOM structure for small inline buttons were similar as
above, just with a few different classes, and no spinner.

For `inline-base-button` based buttons (:tada:), the general DOM
structure now required is this:

```html
<button class="ffe-inline button ffe-inline-button--[button-type]">
  <img class="ffe-inline-button__icon ffe-inline-button__icon--left" alt="left icon" />
  <span class="ffe-inline-button__label">Click me</span>
  <img class="ffe-inline-button__icon ffe-inline-button__icon--right" alt="right icon" />
</button>
```

The icons are optional if not needed. Please refer to the React
implementation for further details on which aria-tags are expected to be
passed where, as well as which modifier flags are available.

selbekk committed:
feat(ffe-buttons-react): Simplify implementation

BREAKING CHANGE! This commit re-implements the ffe-buttons-react
implementation so that it's easier to maintain for future developers.

A missing implementation, `TaskButton`, is also added in this commit.

The opportunity was taken to do some minor API changes as well, since
the prior implementation had some deprecated fields that could be
removed.

With this change, buttons are now much more similar, and follows
(mostly) the same API across button types. There are some differences
between them (some does not support loading indicators, while others
might not support right or left icons. Please refer to the relevant
documentation to learn more about the new APIs.

**Changes in props:**

- `action` prop is removed.
- `simpleContent` prop is removed (only used internally)
- `label` prop is removed. Use `children` instead
- `isTabbable` prop is removed. Use `tabIndex` instead
- `disableButton` prop is renamed to its DOM cousin - `disabled`

**Other changes:**
All buttons support `leftIcon` and `rightIcon` props now - with the
following exceptions:

- `BackButton` does not support any icons by design
- `InlineExpandButton` does not support `rightIcon` by design
- `ShortcutButton` does not support `rightIcon` by design
- `TaskButton` does not support `rightIcon` by design

All buttons now support the `element` prop, which lets you override what
component or DOM element is used to render the button.

wkillerud commented on packages/ffe-buttons-react/src/ExpandButton.js:
```diff
@@ -1,37 +1,49 @@
+                { 'ffe-button--expanded': isExpanded },
+                className,
+            )}
+            leftIcon={isExpanded && <KryssIkon />}
```
If I provide a `leftIcon` I would expect it to apply only to the label in its unexpanded state. Now it will override the `KryssIkon` when `isExpanded` is `true`. `KryssIcon` should perhaps instead be rendered instead of `children`?

That being said, `ExpandButton` should probably not build on `BaseButton` here in React IMO. We don't have an `isLoading` state, and so rendering the spinner and its wrapper seems a bit pointless.

wkillerud replied:
`InlineBaseButton` would be a better parent I think :) Only, this isn't necessarily inline. Hmm..

selbekk replied:
Some good points! 
The reason why this abstraction was used is because the `ffe-button--expand` extends the look and feel of the `ffe-button--secondary`. It shares the same common less-class as well (`ffe-button`), and it follows a lot of the same usage patterns.

As you might have noticed, I added more full-featured prop types to each of the components than they had previously. This is to specify the supported API better than what the previous implementation did. I considered destructuring waay unsupported properties so that misuse would be even harder, but decided against it. I believe undocumented usage should be discouraged, not defensively coded against.

That being said, I'll try to implement a custom component that tries to solve this a bit better.

wkillerud commented on packages/ffe-buttons/less/base-button.less:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    &--action&--ghost {
+        &,
```
Unintended, or should the default and focus state look the same?

selbekk replied:
According to the [original implementation](https://github.com/SpareBank1/designsystem/blob/master/packages/ffe-buttons/less/action-button.less#L16-L21), this is how it should look.

wkillerud reviewed (changes requested):
I'd love the commit messages even more if they had a "before" example as well - how does the markup look today and what are the key changes? (no modifiers inside the button, no label wrapper, but a wrapper for the spinner instead).

I'd also love it if the commit message for the breaking change in React-land mentions the most important breaking changes instead of pointing to the .md immediately.

selbekk commented:
Okay, so I did a few changes:

- Improved the commit logs for easier migration
- Simplified the spinner implementation, now it only needs a single DOM element
- Renamed prop `buttonRef` to `innerRef` (which is a bit easier to standardize across the codebase)
- Re-added `ariaLoadingMessage` as a documented prop, but with a sensible default.
- A few documentation improvements

henrikhermansen reviewed (approved):
Ser veldig bra ut :1st_place_medal:

selbekk commented:
I would really love a review from one of the core maintainers (@kwltrs @antidecaf) on this before it's merged. The changes in the LESS API is big and breaking, so I would love to have most people on board 👯‍♂️

antidecaf reviewed (approved):
LGTM!

selbekk committed:
Merge branch 'master' into 12-reimplement-buttons

selbekk commented:
@kwltrs or @antidecaf  do you mind merging this whenever you're ready to release it? Since it's a complete rewrite, I think it's beneficial if we try to release it as soon as possible after a merge to master.

selbekk committed:
Merge branch 'master' into 12-reimplement-buttons

kwltrs committed:
Merge branch 'master' into 12-reimplement-buttons

wkillerud committed:
Merge branch 'master' into 12-reimplement-buttons

kwltrs committed:
Merge branch 'master' into 12-reimplement-buttons
